### PR TITLE
Make the API return a copy more consistently

### PR DESCRIPTION
**Pull request by pgovind**
Address feedback from https://github.com/dotnet/corefxlab/issues/2734 and other functions I found that could use an `inPlace` option.

### CODE REVIEW

**Comment by eerhardt** [comment]
Tests appear to be failing.

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 1]
```diff
@@ -245,25 +245,39 @@ public void InsertAndRemoveColumnTests()
         public void TestBinaryOperations()
         {
             DataFrame df = MakeDataFrameWithTwoColumns(12);
-            // Binary ops always return a copy
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: It might make sense to leave this comment here

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 2]
```diff
@@ -276,9 +290,24 @@ public void TestBinaryOperations()
             Assert.Equal(false, df.GreaterThan(listOfInts)[5, 0]);
             Assert.Equal(false, df.LessThan(5)[5, 0]);
             Assert.Equal(false, df.LessThan(listOfInts)[5, 0]);
-
-            // The original DF is untouched
-            Assert.Equal(0, df[0, 0]);
+            // The following binary ops are in place
+            Assert.Equal(5, df.Add(5, true)[0, 0]);
```
Comment (eerhardt) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: A nice, readable way of coding with booleans is to use named parameters:

```C#
Assert.Equal(5, df.Add(5, inPlace: true)[0, 0]);
```

It makes it really easy for a new person reading the code understand what `true` means here.

File: src/Microsoft.Data/PrimitiveColumn.cs [hunk 3]
```diff
@@ -527,11 +527,14 @@ private PrimitiveColumn<T> _Filter(T lower, T upper)
             return ret;
         }
 
-        private PrimitiveColumn<T> _Clip(T lower, T upper)
+        private PrimitiveColumn<T> _Clip(T lower, T upper, bool inPlace)
         {
-            PrimitiveColumn<T> ret = Clone() as PrimitiveColumn<T>;
+            PrimitiveColumn<T> ret = this;
+            if (!inPlace)
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumn.cs]:
```C#
PrimitiveColumn<T> ret =inPlace ? this : Clone();
```

?

File: src/Microsoft.Data/DataFrame.cs [hunk 4]
```diff
@@ -378,51 +378,62 @@ public DataFrame Sort(string columnName, bool ascending = true)
...
         /// <summary>
         /// Adds a prefix to the column names
         /// </summary>
-        public DataFrame AddPrefix(string prefix)
+        public DataFrame AddPrefix(string prefix, bool inPlace = false)
         {
-            for (int i = 0; i < ColumnCount; i++)
+            DataFrame df = this;
+            if (!inPlace)
             {
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.cs]: It would be nice to consistently brace single line ifs or not. Above you don't use braces, but here you do.

(my personal preference is to always use braces.)

File: src/Microsoft.Data/DataFrame.cs [hunk 5]
```diff
@@ -500,27 +511,33 @@ public DataFrame DropNulls(DropNullOptions options = DropNullOptions.Any)
...
-            DataFrame ret = new DataFrame();
-            for (int i = 0; i < ColumnCount; i++)
+            DataFrame ret = this;
+            if (!inPlace)
+                ret = Clone();
+
+            for (int i = 0; i < ret.ColumnCount; i++)
             {
-                ret.InsertColumn(i, Column(i).FillNulls(value));
+                ret.Column(i).FillNulls(value, true);
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.cs]: Similar comment here about naming boolean parameters for easy reading.  `inPlace: true`.

File: src/Microsoft.Data/ArrowStringColumn.cs [hunk 6]
```diff
@@ -198,7 +198,7 @@ private void Append(ReadOnlySpan<byte> value)
                 mutableDataBuffer.Length += value.Length;
                 mutableOffsetsBuffer.Append(mutableOffsetsBuffer[mutableOffsetsBuffer.Length - 1] + value.Length);
             }
-            SetValidityBit(Length - 1, value == null ? true : false);
+            SetValidityBit(Length - 1, value == default ? false : true);
```
Comment (eerhardt) [inline src/Microsoft.Data/ArrowStringColumn.cs]: >value [](start = 39, length = 5)

Could simply be:

```C#
SetValidityBit(Length - 1, value != default);
```

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs [hunk 7]
```diff
@@ -14,419 +14,539 @@ public partial class DataFrame
...
+        public DataFrame Add<T>(IReadOnlyList<T> values, bool inPlace = false)
             where T : unmanaged
         {
             if (values.Count != ColumnCount)
             {
-                throw new ArgumentException($"values.Count {values.Count} must match the number of columns in the table", nameof(values));
+                throw new ArgumentException(Strings.MismatchedColumnLengths, nameof(values));
             }
-            var newDataFrame = new DataFrame();
+            DataFrame retDataFrame = this;
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.BinaryOperations.cs]: Similar comment here about using a tertiary operator:

```C#
DataFrame retDataFrame = inPlace ? this : new DataFrame();
```

File: src/Microsoft.Data/ArrowStringColumn.cs [hunk 8]
```diff
@@ -94,20 +94,20 @@ private void SetValidityBit(long index, bool value)
             if (value)
             {
                 newBitMap = (byte)(curBitMap | (byte)(1 << (indexInBuffer & 7))); //bit hack for index % 8
-                if ((curBitMap >> (indexInBuffer & 7) & 1) == 0 && indexInBuffer < Length && NullCount > 0)
+                if (((curBitMap >> (indexInBuffer & 7)) & 1) == 0 && indexInBuffer < Length - 1 && NullCount > 0)
```
Comment (pgovind) [inline src/Microsoft.Data/ArrowStringColumn.cs]: Not an unrelated bug fix. The "inPlace" tests exposed this.

**Comment by pgovind** [comment]
This fixes https://github.com/dotnet/corefxlab/issues/2739.

File: src/Microsoft.Data/DataFrame.BinaryOperations.cs [hunk 9]
```diff
@@ -14,419 +14,503 @@ public partial class DataFrame
...
         #region Binary Operations
 
-        public DataFrame Add<T>(IReadOnlyList<T> values)
+        public DataFrame Add<T>(IReadOnlyList<T> values, bool inPlace = false)
             where T : unmanaged
         {
             if (values.Count != ColumnCount)
             {
-                throw new ArgumentException($"values.Count {values.Count} must match the number of columns in the table", nameof(values));
+                throw new ArgumentException(Strings.MismatchedColumnLengths, nameof(values));
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.BinaryOperations.cs]: (minor) it may be nice to log both `values.Count` and `ColumnCount` here in the exception message. That way when this occurs people can see the values without having to debug it.

Comment (pgovind) [reply src/Microsoft.Data/DataFrame.BinaryOperations.cs]: I'll do this slowly in the other PRs. I think I need to refactor and possibly split this message.
